parser/impl/parser.cpp

```diff
@@ -1,18 +1,59 @@
-#include "parser.hpp"
-#include "token.hpp"
+#include <algorithm>
 #include <memory>
+#include "token.hpp"
+
+#include "parser.hpp"
 
 namespace parser {
 
 using lexer::Token;
 using lexer::TokenType;
 
-static std::array<TokenType, 3> primitives{
+static const std::array<TokenType, 3> primitives{
     TokenType::IntegerType,
     TokenType::RealType,
     TokenType::Boolean,
 };
 
+namespace util {
+
+
+inline bool IsPrimitiveType(TokenType type) {
+    return std::count(primitives.begin(), primitives.end(), type) != 0;
+}
+
+template <typename Container, typename T>
+inline auto Contains(const Container& data,
+                     const T& value)
+                    -> decltype(std::end(data), true) {
+    return std::end(data) != std::find(std::begin(data), std::end(data), value);
+}
+
+std::string GenExpectMessage(const std::vector<TokenType>& types) {
+    if (types.size() == 0) {
+        return "expected nothing";
+    }
+
+    if (types.size() == 1) {
+        return fmt::format("expected {}", lexer::to_string(types[0]));
+    }
+
+    if (types.size() == 2) {
+        return fmt::format("expected {} or {}", 
+                        lexer::to_string(types[0]), lexer::to_string(types[1]));
+    }
+
+    std::string msg = "expected ";
```

Comment:
I decided to make all messages with the lower case latter, and omit periods at the ends of the sentences too. It looks a bit better (my opinion), but most importantly it will allow us to sustain syntactically bugless sentences easier, since it's all the same.